include/triton/backend/device_memory_tracker.h

```
@@ -0,0 +1,287 @@
+  uint32_t cuda_array_len_;
+
+  // only set to false if somehow the CUPTI activity occurs on index out of
+  // range. In that case, user should invalidate the whole tracker.
+  bool good_record_;
+} TRITONBACKEND_CuptiTracker;
+
+}
+
+class DeviceMemoryTracker {
+ public:
+  struct MemoryUsage {
+    MemoryUsage()
+    {
+      cuda_byte_size_.resize(CudaDeviceCount(), 0);
+
+      cupti_tracker_.system_byte_size_ = system_byte_size_.data();
+      cupti_tracker_.pinned_byte_size_ = pinned_byte_size_.data();
+      cupti_tracker_.cuda_byte_size_ = cuda_byte_size_.data();
+      cupti_tracker_.system_array_len_ = system_byte_size_.size();
+      cupti_tracker_.pinned_array_len_ = pinned_byte_size_.size();
+      cupti_tracker_.cuda_array_len_ = cuda_byte_size_.size();
+      cupti_tracker_.good_record_ = true;
+    }
+
+    ~MemoryUsage()
+    {
+      // Make sure all C struct reference are dropped before clearing.
+      if (tracked_) {
+        UntrackThreadMemoryUsage(this);
+      }
+    }
+
+    // Disable copy and assign to better manage C struct lifecycle
+    MemoryUsage(const MemoryUsage&) = delete;
+    void operator=(const MemoryUsage&) = delete;
+
+    // merge record from another MemoryUsage object
+    MemoryUsage& operator+=(const MemoryUsage& rhs)
+    {
+      std::transform(rhs.system_byte_size_.begin(), rhs.system_byte_size_.end(), system_byte_size_.begin(), system_byte_size_.begin(),
+      std::plus<int64_t>());
+      std::transform(rhs.pinned_byte_size_.begin(), rhs.pinned_byte_size_.end(), pinned_byte_size_.begin(), pinned_byte_size_.begin(),
+      std::plus<int64_t>());
+      std::transform(rhs.cuda_byte_size_.begin(), rhs.cuda_byte_size_.end(), cuda_byte_size_.begin(), cuda_byte_size_.begin(),
+      std::plus<int64_t>());
+      return *this;
+    }
+
+    TRITONSERVER_Error* SerializeToBufferAttributes(
+      TRITONSERVER_BufferAttributes** usage, int32_t* usage_size)
+    {
+      int32_t usage_idx = 0;
```

should this fail if the cupti record is invalid?